crates/astria-grpc-mock/src/mounted_mock.rs

```diff
@@ -0,0 +1,202 @@
+        {
+            writeln!(buffer, "{body}")
+        } else {
+            writeln!(buffer, "<Could not map the gRPC body to JSON>")
+        }
+    }
+}
+
+impl From<(Request<AnyMessage>, MockResponse)> for BadResponse {
+    fn from(value: (Request<AnyMessage>, MockResponse)) -> Self {
+        Self {
+            request: value.0,
+            mock_response: value.1,
+        }
+    }
+}
+
+impl Clone for BadResponse {
+    fn clone(&self) -> Self {
+        Self {
+            request: clone_request(&self.request),
+            mock_response: self.mock_response.clone(),
+        }
+    }
+}
+
+pub(crate) struct MountedMock {
+    inner: Mock,
+    position_in_set: usize,
+    notify: Arc<(Notify, AtomicBool)>,
+    successful_response: Vec<(Request<AnyMessage>, ResponseResult)>,
+    bad_responses: Vec<BadResponse>,
+}
+
+impl MountedMock {
+    pub(crate) fn match_and_respond<U: 'static>(
+        &mut self,
+        rpc: &'static str,
+        request: &Request<AnyMessage>,
+    ) -> MockResult<U> {
+        if self.inner.rpc != rpc
+            || !self
+                .inner
+                .matchers
+                .iter()
+                .all(|matcher| matcher.matches(request))
+        {
+            return MockResult::NoMatch;
+        }
+
+        let response = match self.inner.response.respond(request) {
+            Err(status) => {
+                self.successful_response
+                    .push((clone_request(request), Err(status.clone())));
+                Ok(Err(status))
+            }
+            Ok(mock_response) => {
+                let (metadata, erased_message, extensions) =
+                    clone_response(&mock_response.inner).into_parts();
+                if let Ok(message) = erased_message.clone_box().into_any().downcast::<U>() {
+                    let rsp = tonic::Response::from_parts(metadata, *message, extensions);
+                    self.successful_response
+                        .push((clone_request(request), Ok(mock_response)));
+                    Ok(Ok(rsp))
+                } else {
+                    let actual = mock_response.type_name;
+                    self.bad_responses
+                        .push((clone_request(request), mock_response).into());
+                    let expected = std::any::type_name::<U>();
+                    #[rustfmt::skip]
+                        let msg = format!(
+                            "failed downcasting mock response to concrete type:\n\
+                             required type of gRPC response: `{expected}`\n\
+                             type of mock response: `{actual}`\n\
+                             JSON serialization:\n\
+                             {}",
+                            serde_json::to_string_pretty(erased_message.as_serialize())
+                                    .expect("can map registered protobuf response to json")
+                        );
+                    Err(tonic::Status::internal(msg))
```

Comment:
when would this error case happen? if the server is implemented to return type X but the mock response implements type Y?

Reply:
yes, exactly. On line 134 we try and downcast the response to to `U`. If we are unable to do that the mock server returns an error (because we need to return something).